pageserver/src/http/routes.rs

```diff
@@ -235,6 +239,36 @@ async fn timeline_detail_handler(request: Request<Body>) -> Result<Response<Body
     Ok(json_response(StatusCode::OK, response_data)?)
 }
 
+async fn timeline_attach_handler(request: Request<Body>) -> Result<Response<Body>, ApiError> {
+    let tenant_id: ZTenantId = parse_request_param(&request, "tenant_id")?;
+    check_permission(&request, Some(tenant_id))?;
+
+    let timeline_id: ZTimelineId = parse_request_param(&request, "timeline_id")?;
+
+    let response_data = tokio::task::spawn_blocking(move || {
+        let _enter =
+            info_span!("timeline_attach_handler", tenant = %tenant_id, timeline = %timeline_id)
+                .entered();
+        let repo = tenant_mgr::get_repository_for_tenant(tenant_id)?;
+        match repo.get_timeline(timeline_id)? {
+            RepositoryTimeline::Local(_) => {
+                bail!("Timeline with id {} is already local", timeline_id)
+            }
+            RepositoryTimeline::Remote(_) => {
+                // FIXME this is a bit hacky because there is local_or_schedule_download on LayeredTimelineEntry, but not on RepositoryTimeline
+                // Also it is a good idea to change timeline status to AwaitsDownload (TODO probably incapsulate it somewhere in the repo)
+                repo.set_timeline_state(timeline_id, TimelineSyncState::AwaitsDownload)?;
+                schedule_timeline_download(tenant_id, timeline_id);
```

Comment:
I know this is a bit of a weird naming now, but `repo.get_timeline` now, being implemented by the `LayeredRepository`, always attempts to download (if not downloaded) and read from disk and initialize (if not intialized) before returning a timeline, so no need to schedule anything else on top, I assume?

Impl details:
https://github.com/zenithdb/zenith/blob/5cff7d1de9862d8a826c4454d997a193ffa26ba6/pageserver/src/layered_repository.rs#L140


> is a good idea to change timeline status to AwaitsDownload

Yes, good idea, yet not super strictly necessary now: we already return a techically correct `CloudOnly` one on the `LayeredRepository` impl: https://github.com/zenithdb/zenith/blob/5cff7d1de9862d8a826c4454d997a193ffa26ba6/pageserver/src/layered_repository.rs#L326

and it should be better when we sort out repeated downloads for already downloaded timeline.

Reply:
Hm, I think on demand nature of get_timeline is a bit awkward, for example when you want to inspect the state without changing it. IMO decision to download missing files (or initialize the local repo) should me made by the calling code. If this is the most common pattern we can have multiple methods, like get_or_download or something.

Also there is a branch create api, so we should know upfront when new timeline is being initialized, so maybe we can  avoid this confusion by throwing an error on missing timeline instead of downloading/initialization. What do you think?

Reply:
Ideally, yes, that would be great.
Realistically, we have to go over each such invocation now and assert its behavior, possibly amending the code around.
I think it's better to discuss it a bit first and then do the changes in a separate PR, if agreed on.

Reply:
See https://github.com/zenithdb/zenith/issues/997